--- control
+++ control
@@ -4,7 +4,7 @@
 Maintainer: Debian PostgreSQL Maintainers <team+postgresql@tracker.debian.org>
 Uploaders: Adrian Vondendriesch <adrian.vondendriesch@credativ.de>
 Build-Depends: debhelper (>= 9),
-               libperl-dev,
+               perl,
                libpq-dev,
                libssl-dev,
                libxml-checker-perl,
@@ -18,10 +18,8 @@
 
 Package: pgbackrest
 Architecture: any
-Depends: perl,
-         postgresql-common,
+Depends: postgresql-common,
          ${misc:Depends},
-         ${perl:Depends},
          ${shlibs:Depends}
 Suggests: pgbackrest-doc
 Description: Reliable PostgreSQL Backup & Restore
--- copyright
+++ copyright
@@ -3,8 +3,8 @@
 Source: https://github.com/pgbackrest/pgbackrest
 
 Files: *
-Copyright: 2015-2016 The PostgreSQL Global Development Group
-           2013-2016 David Steele
+Copyright: 2015-2020 The PostgreSQL Global Development Group
+           2013-2020 David Steele
 License: MIT
 
 Files: debian/*
