--- rules
+++ rules
@@ -23,10 +23,9 @@
 	  -t pgbackrest \
 	  -s 1 \
 	  ${MANTEMPLATE} > ${CURDIR}/doc/output/man/pgbackrest.1
-	# The Makefile doesn't accepts *FLAGS as environment variables.
-	# However, it offers a method to pass extra arguments via *EXTRA
-	# variable.
-	make -C $(CURDIR)/src/ CEXTRA="$(CFLAGS)" LDEXTRA="$(LDFLAGS)"
+
+	cd $(CURDIR)/src && ./configure
+	make -C $(CURDIR)/src
 	dh_auto_build
 
 override_dh_auto_clean:
