--- pgbackrest.spec
+++ pgbackrest.spec
@@ -8,16 +8,7 @@
 Url:		http://www.pgbackrest.org/
 Source0:	https://github.com/pgbackrest/pgbackrest/archive/release/%{version}.tar.gz
 Source1:	pgbackrest-conf.patch
-Requires:	perl-XML-LibXML perl-IO-Socket-SSL
-%if 0%{?rhel} && 0%{?rhel} <= 6
-Requires:	perl-parent perl-JSON perl-Time-HiRes
-%else
-Requires:	perl-JSON-PP
-%endif
-Requires:	perl-Digest-SHA perl-DBD-Pg perl-Time-HiRes zlib
-Requires:	perl(:MODULE_COMPAT_%(eval "`%{__perl} -V:version`"; echo $version))
-BuildRequires:	openssl-devel zlib-devel perl-ExtUtils-Embed
-BuildRequires:	postgresql%{pgmajorversion}-devel
+BuildRequires:	openssl-devel zlib-devel postgresql%{pgmajorversion}-devel
 Requires:	postgresql-libs
 
 
