--- pgbackrest.spec
+++ pgbackrest.spec
@@ -28,11 +28,6 @@
 %setup -q -n %{name}-release-%{version}
 
 %build
-pushd libc
-perl Makefile.PL
-%{__make}
-popd
-
 pushd src
 %{__make}
 popd
@@ -44,12 +39,6 @@
 %{__install} -D -d -m 0700 %{buildroot}/var/spool/%{name}
 %{__install} -D -d -m 0755 %{buildroot}%{_sysconfdir}
 %{__install} %{SOURCE1} %{buildroot}/%{_sysconfdir}/%{name}.conf
-%{__cp} -a lib/* %{buildroot}%{perl_vendorlib}/
-%{__mkdir} -p %{buildroot}%{perl_vendorarch}/auto/pgBackRest/LibC
-%{__cp} -a libc/blib/arch/auto/pgBackRest/LibC/* %{buildroot}%{perl_vendorarch}/auto/pgBackRest/LibC
-%{__cp} -a libc/blib/lib/auto/pgBackRest/LibC/* %{buildroot}%{perl_vendorarch}/auto/pgBackRest/LibC
-%{__mkdir} -p %{buildroot}%{perl_vendorarch}/pgBackRest
-%{__cp} -a libc/blib/lib/pgBackRest/* %{buildroot}%{perl_vendorarch}/pgBackRest
 %{__cp} -a src/%{name} %{buildroot}%{_bindir}/%{name}
 
 %clean
@@ -64,9 +53,6 @@
 %endif
 %{_bindir}/%{name}
 %config(noreplace) %attr (644,root,root) %{_sysconfdir}/%{name}.conf
-%{perl_vendorlib}/pgBackRest/
-%{perl_vendorarch}/pgBackRest/
-%{perl_vendorarch}/auto/pgBackRest/LibC
 %attr(-,postgres,postgres) /var/log/%{name}
 %attr(-,postgres,postgres) %{_sharedstatedir}/%{name}
 %attr(-,postgres,postgres) /var/spool/%{name}
