--- pgbackrest.spec
+++ pgbackrest.spec
@@ -8,8 +8,7 @@
 Url:		http://www.pgbackrest.org/
 Source0:	https://github.com/pgbackrest/pgbackrest/archive/release/%{version}.tar.gz
 Source1:	pgbackrest-conf.patch
-Patch0:		pgbackrest-libxmlinclude.patch
-BuildRequires:	openssl-devel zlib-devel postgresql%{pgmajorversion}-devel
+BuildRequires:	openssl-devel zlib-devel postgresql%{pgmajorversion}-devel lz4-devel
 Requires:	postgresql-libs
 
 
@@ -26,7 +25,6 @@
 
 %prep
 %setup -q -n %{name}-release-%{version}
-%patch0 -p0
 
 %build
 pushd src
